File: src/algorithms/cpp/bucketSort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

vector<int> bucketSort(vector<int> array)
{
    int bucketSize = 10;
    int maxValue = array[0];
    int minValue = array[0];

    for (int value : array)
    {
        if (value > maxValue)
        {
            maxValue = value;
        }
        if (value < minValue)
        {
            minValue = value;
        }
    }

    int bucketCount = floor((maxValue - minValue) / bucketSize) + 1;
    vector<vector<int>> buckets(bucketCount);

    size_t n = array.size();

    for (size_t i = 0; i < n; i++)
    {
        int index = floor((array[i] - minValue) / bucketSize);
        buckets[index].push_back(array[i]);
    }

    array.clear(); // Limpiar el arreglo original

    for (vector<int> bucket : buckets)
    {
        sort(bucket.begin(), bucket.end());
        array.insert(array.end(), bucket.begin(), bucket.end());
    }

    return array;
}

File: src/algorithms/go/bucketSort.go
package main

import (
	"math"
)

func bucketSort(array []int) []int {
	bucketSize := 10
	maxValue := array[0]
	minValue := array[0]

	for _, value := range array {
		if value > maxValue {
			maxValue = value
		}
		if value < minValue {
			minValue = value
		}
	}

	bucketCount := int(math.Floor(float64(maxValue-minValue)/float64(bucketSize))) + 1
	buckets := make([][]int, bucketCount)

	for i := 0; i < len(array); i++ {
		index := int(math.Floor(float64(array[i]-minValue) / float64(bucketSize)))
		buckets[index] = append(buckets[index], array[i])
	}

	array = array[:0] // Limpiar el arreglo original

	for _, bucket := range buckets {
		insertionSort(bucket)
		array = append(array, bucket...)
	}

	return array
}
